File: 雷达仿真/FMCW_2D.m
% https://mp.weixin.qq.com/s?__biz=Mzk0NTQzNDQyMw==&mid=2247503888&idx=5&sn=91128cb13ac6630210d6b8b6499d4e64&chksm=c2ed302e9bac292547e68672237cf6c0cd225ce1341b7b3729cbe9a3b527092397fcee36e16f&mpshare=1&scene=1&srcid=0418eJ3yhGBfCih3fAoII6m8&sharer_shareinfo=8b359d0170425ccb2bde8c4af91d4c4b&sharer_shareinfo_first=8b359d0170425ccb2bde8c4af91d4c4b&exportkey=n_ChQIAhIQWRiY%2Fy5a7x2s3ez1NEpK9BKfAgIE97dBBAEAAAAAANFbCZdm9S0AAAAOpnltbLcz9gKNyK89dVj0UcD%2B%2BkUEbFZE1wQghzG8A%2BZdsriQPSbPcdbYuPLqEhN%2FArUqtVYA9UJt13wnxHKC5VHMjtIKs7cqkz%2B2qFltxVtNFZF9AUFS4GeXdkf%2FUeCuM2OHh9V7mT8YNqx4WRWYgldh0WwTqbyPsdH4uC8d9Y0lDGbW0KoEmO8vOYfVscVQvL0fR0vKSv4nsyqnvRtpAfOTCXJtRivCdwBs72laJHTwhqGxnaMQo6ojdDCh9RDbKFcag2It99L0qFuxmSFo%2FNOh1J1UL6m7eeAPz8Q%2BG6pOcZtnIdaxOcHKH6fHY30n1fGzd%2BA9YifyB%2B8kAQkcY858ih9sPFMF&acctmode=0&pass_ticket=gOtBYAlXM34c%2FAP6uuptKu57VPg8uNOrz0qOvCVuc9ipvqAOJQ6bUnW0DZyuQOnO&wx_header=0#rd
clc;
clear all;
close all;
%It is for FMCW Radars
%% Define Radar parameters
radar_parameters.max_range = 25; % Maximum Unambiguous Range in meters
radar_parameters.max_velocity =20; % Maximum Velocity in m/s
radar_parameters.L = 128; % Number of Chirps
radar_parameters.N = 256; % Number of Samples per Chirp
radar_parameters.B = 1500e6; % Sweep Bandwidth in Hz
radar_parameters.f_start = 78e9; % Carrier Frequency in Hz
radar_parameters.T= 50e-6; % Chirp Duration in seconds
radar_parameters.TRRI = 60e-6;% Ramp Repetion Interval
radar_parameters.f_s = 5e6; % ADC Sampling Rate in Hz
radar_parameters.T_s = 1/radar_parameters.f_s;
radar_parameters.S = radar_parameters.B/radar_parameters.T; % Ramp Slope in Hz/s
target_range=5;
target_velocity=10;
c = 3e8; % Speed of light in m/s
%% Call the function adcDataGenerate
Rx = adcDataGenerate(radar_parameters,target_range,target_velocity);
%% Noise Generation
signal_power = 1; % Signal power
SNR_dB = 20; % SNR in dB

% Convert SNR from dB to linear scale
SNR_linear = db2pow(SNR_dB);

% Calculate noise power
noise_power = signal_power / SNR_linear;

% Calculate standard deviation from noise power(mean=0)
sigma = sqrt(noise_power / 2);

% Generate real and imaginary parts
real_part = sigma * randn(radar_parameters.N, radar_parameters.L); 
imaginary_part = sigma * randn(radar_parameters.N, radar_parameters.L); 

% Combining real and imaginary parts
Noise=complex(real_part,imaginary_part);

X=Rx+Noise;
%% DFT Signal Processing
F_n = (1/sqrt(radar_parameters.N))*dftmtx(radar_parameters.N); %DFT matrix with N*N
F_l = (1/sqrt(radar_parameters.L))*dftmtx(radar_parameters.L); %DFT matrix with L*L
X_2d = (F_n)*X*(transpose(F_l)); %The expression for finding the 2D DFT 

% Range Resolution
del_R = (radar_parameters.T * c) / (2 * radar_parameters.B*radar_parameters.T_s*radar_parameters.N);

% Velocity Resolution
del_v = c / (2 * radar_parameters.f_start * radar_parameters.TRRI * radar_parameters.L);

%Range and velocity limits
R_axis = 0:del_R:((radar_parameters.N/2)-1)* del_R ;
v_axis = -del_v*(radar_parameters.L/2) :del_v:del_v*((radar_parameters.L/2) - 1);
[v_grid,R_grid] = meshgrid(v_axis,R_axis);

%Accessing only N/2 DFT coeffients
X_2d = X_2d(1:radar_parameters.N/2, :);

% Define the limits for column swapping
first_half = 1:radar_parameters.L/2;
second_half = (radar_parameters.L/2)+1:radar_parameters.L;

% Swap columns with defined limits
X_2d(:, [first_half, second_half]) = X_2d(:, [second_half, first_half]);

% Frequency grids for plotting
k_axis = 0:(radar_parameters.N/2)-1;
p_axis = -radar_parameters.L/2 :(radar_parameters.L/2) - 1;
[p_grid, k_grid] = meshgrid(p_axis, k_axis);

% Plot magnitude of 2D DFT
figure;
surf(p_grid, k_grid, abs(X_2d), 'EdgeColor', 'none');
xlabel('p axis');
ylabel('k axis');
zlabel('Magnitude');
title('3D Plot of 2D DFT Magnitude in frequency indices');

%% Plot magnitude of 2D DFT
figure;
surf(v_grid, R_grid, abs(X_2d), 'EdgeColor', 'none');
xlabel('Velocity (m/s)');
ylabel('Range (m)');
zlabel('Magnitude');
title('3D Plot of 2D DFT Magnitude in Range and Velocity');

R_max=(radar_parameters.f_s*c)/(2*radar_parameters.S);
disp(R_max);
v_max=c/(4*radar_parameters.T*radar_parameters.f_start);
disp(v_max);

% Function adcDataGenerate that simulates the data from the given structue radarparamets
function Rx = adcDataGenerate(radar_parameters, target_range, target_velocity)
    c = 3e8;                   %  Speed of light in m/s
    N = radar_parameters.N;
    L = radar_parameters.L; 
    T = radar_parameters.T;
    B = radar_parameters.B;
    max_range = radar_parameters.max_range;
    f_start = radar_parameters.f_start;
    TRRI = radar_parameters.TRRI;
    S = radar_parameters.S;
    f_s = radar_parameters.f_s;
    T_s = radar_parameters.T_s;              % Sampling Period
    R = target_range;
    v = target_velocity; 
    max_prop_time = 3 *max_range/c;        % Maximum delay
    for Chi_idx=1:L                        % Chirp index for L Chirps
        for Sam_idx=1:N                    % Sample index for N Samples
            t = max_prop_time +(Sam_idx-1)*T_s;
            phase_Tx(Sam_idx,Chi_idx) = 2*pi*f_start*t+pi*S*t^2;
            % Phase  of the received Signal
            tau = 2*(R+v*t+v*TRRI*(Chi_idx-1))/c;
            phase_Rx(Sam_idx,Chi_idx) = 2 * pi * f_start *(t-tau) + pi * S * (t-tau)^2;
            % Phase of the mixed Signal/IF signal 
            phase_IF(Sam_idx,Chi_idx)=phase_Tx(Sam_idx,Chi_idx)- phase_Rx(Sam_idx,Chi_idx); 
            % The Mixed Signal/ IF signal
            Rx(Sam_idx,Chi_idx)= complex(cos(phase_IF(Sam_idx,Chi_idx)),sin(phase_IF(Sam_idx,Chi_idx))); % Complex command
        end
    end 
end

File: 雷达仿真/FMCW_2D.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 18 19:43:36 2025

@author: jack

单目标，直接利用解析形式的差频信号，利用FFT矩阵进行结算.

"""

import numpy as np
import matplotlib.pyplot as plt
import scipy

# 全局设置字体大小
# plt.rcParams["font.family"] = "Times New Roman"
plt.rcParams["font.family"] = "SimSun"
plt.rcParams['font.size'] = 18               # 设置全局字体大小
plt.rcParams['axes.titlesize'] = 18          # 设置坐标轴标题字体大小
plt.rcParams['axes.labelsize'] = 18          # 设置坐标轴标签字体大小
plt.rcParams['xtick.labelsize'] = 18         # 设置 x 轴刻度字体大小
plt.rcParams['ytick.labelsize'] = 18         # 设置 y 轴刻度字体大小
plt.rcParams['axes.unicode_minus'] = False   # 用来显示负号
plt.rcParams["figure.figsize"] = [8, 6]      # 调整生成的图表最大尺寸
# plt.rcParams['figure.dpi'] = 300           # 每英寸点数
plt.rcParams['lines.linestyle'] = '-'
plt.rcParams['lines.linewidth'] = 2          # 线条宽度
plt.rcParams['lines.color'] = 'blue'
plt.rcParams['lines.markersize'] = 6         # 标记大小
# plt.rcParams['figure.facecolor'] = 'lightgrey'   # 设置图形背景色为浅灰色
plt.rcParams['figure.facecolor'] = 'white'         # 设置图形背景色为浅灰色
plt.rcParams['axes.edgecolor'] = 'black'           # 设置坐标轴边框颜色为黑色
plt.rcParams['legend.fontsize'] = 18


def FFTmatrix(N):
     M = np.zeros((N, N), dtype = complex)
     tmp = np.arange(N)
     for n in range(N):
         M[n, :] =  np.exp(-1j*2*np.pi*n*tmp/N)
     return M

c = 3e8
Rmax  = 25     # Maximum Unambiguous Range in meters
Vmax = 20      # Maximum Velocity in m/s
Nchirps = 128  # Number of Chirps
Ns = 256       # Number of Samples per Chirp
B = 1.5e9      # Sweep Bandwidth in Hz
f0 = 78e9      # Carrier Frequency in Hz
Tc = 50e-6     # Chirp Duration in seconds
TRRI = 60e-6   # Ramp Repetion Interval
fs = 5e6       # ADC Sampling Rate in Hz
Ts = 1/fs      #
S = B/Tc

tarR = 5
tarV = 10

R_max = (fs * c)/(2*S)
v_max = c/(4 * Tc * f0)
print(f"tarR = {tarR}, R_max = {R_max}, tarV = {tarV}, v_max = {v_max}")

Rx = np.zeros((Ns, Nchirps), dtype = complex)
max_prop = 3 * Rmax/c

## get IF signal
for nc in range(Nchirps):
    Sam_idx = np.arange(Ns)
    t = max_prop + Sam_idx * Ts
    phase_Tx = 2 * np.pi * f0 * t + np.pi * S * t**2
    tau = 2*(tarR + tarV * t + tarV * TRRI * nc)/c
    phase_Rx = 2 * np.pi * f0 *(t - tau) + np.pi * S * (t-tau)**2
    phase_IF = phase_Tx - phase_Rx
    Rx[:,nc] = np.cos(phase_IF) + 1j * np.sin(phase_IF)

signal_power = np.sum(np.abs(Rx)**2)/Rx.size
SNRdB = 20
noise_pow = signal_power / (10.0**(SNRdB/10.0))

X = Rx + np.sqrt(noise_pow/2) * (np.random.randn(*Rx.shape) + 1j * np.random.randn(*Rx.shape))

## DFT Signal Processing
F_n = (1/np.sqrt(Ns))*FFTmatrix(Ns)              # DFT matrix with N*N
F_l = (1/np.sqrt(Nchirps))*FFTmatrix(Nchirps)    # DFT matrix with L*L
X_2d =  F_n  @ X @ (F_l.T)                       # The expression for finding the 2D DFT

## Range Resolution
del_R = (Tc * c) / (2 * B * Ts * Ns)

## Velocity Resolution
del_v = c / (2 * f0 * TRRI * Nchirps)

## Range and velocity limits
R_axis = np.arange(0, Ns/2 * del_R, del_R)
v_axis = np.arange(-del_v*Nchirps/2, del_v*Nchirps/2, del_v)
v_grid, R_grid = np.meshgrid(v_axis, R_axis)

## Accessing only N/2 DFT coeffients
X_2d = X_2d[:int(Ns/2),:] #   (1:radar_parameters.N/2, :);

## Swap columns with defined limits
tmp = X_2d[:, :int(Nchirps/2)].copy()
X_2d[:, :int(Nchirps/2)] = X_2d[:, int(Nchirps/2):]
X_2d[:, int(Nchirps/2):] = tmp

## Frequency grids for plotting
k_axis = np.arange(Ns/2)
p_axis = np.arange(-Nchirps/2, Nchirps/2, 1)
p_grid, k_grid = np.meshgrid(p_axis, k_axis);

## Plot magnitude of 2D DFT
fig = plt.figure(figsize=(10, 10) )
ax1 = fig.add_subplot(111, projection = '3d')
ax1.plot_surface(p_grid, k_grid, np.abs(X_2d), rstride = 5, cstride = 5, cmap = plt.get_cmap('jet'))
ax1.grid(False)
ax1.set_proj_type('ortho')
ax1.set_xlabel('距离门数', )
ax1.set_ylabel('脉冲数', )
ax1.set_title('Magnitude', )
ax1.view_init(azim = -135, elev = 30)
plt.show()
plt.close()

## Plot magnitude of 2D DFT
fig = plt.figure(figsize=(10, 10) )
ax1 = fig.add_subplot(111, projection = '3d')
ax1.plot_surface(v_grid, R_grid, np.abs(X_2d), rstride = 5, cstride = 5, cmap = plt.get_cmap('jet'))
ax1.grid(False)
ax1.set_proj_type('ortho')
ax1.set_xlabel('Velocity (m/s)', )
ax1.set_ylabel('Range (m)', )
ax1.set_title('Magnitude', )
ax1.view_init(azim = -135, elev = 30)
plt.show()
plt.close()
















































